### DISCUSSION TITLE
Spark-3.1.2-Azure/Synapse-SQL-Almost Works !

### DISCUSSION THREAD

**Discussion by Grudra1** [OpenLineage/OpenLineage#653] [Q&A]
I was trying to get the lineage for a sample job which reads and writes data from Synapse Sql. I know it is not supported for Saprk lineage yet, but since in the documentation, https://github.com/microsoft/sql-spark-connector, 
it says : 
[](https://github.com/microsoft/sql-spark-connector#supported-options)Supported Options

_The Apache Spark Connector for SQL Server and Azure SQL supports the options defined here: [SQL DataSource JDBC](https://spark.apache.org/docs/latest/sql-data-sources-jdbc.html)_

which I take as a hint it is built on top of spark jdbc connector and Spark integration supports jdbc connectors for datasets , I gave it a shot.
Well the result is interesting, because it can put the lineage for reads from Synapse, but while I tried to write the dataframe into a Synapse table, it breaks just while trying to post it on Marquez. The event is properly created by the openlineage integration, just the post fails. Error trace is as given below :

22/04/05 11:04:48 ERROR OpenLineageContext: Could not emit lineage: 
{"eventType":"COMPLETE","eventTime":"2022-04-05T11:24:34.411Z","run":{"runId":"76491cd9-9abd-457c-9068-250999993c9d","facets":{"spark.logicalPlan":{"_producer":"https://github.com/OpenLineage/OpenLineage/tree/0.4.0/integration/spark","_schemaURL":"https://openlineage.io/spec/1-0-2/OpenLineage.json#/$defs/RunFacet","plan":[{"class":"org.apache.spark.sql.execution.datasources.SaveIntoDataSourceCommand","num-children":0,"query":[{"class":"org.apache.spark.sql.catalyst.plans.logical.Project","num-children":1,"projectList":[[{"class":"org.apache.spark.sql.catalyst.expressions.Alias","num-children":1,"child":0,"name":"VENDORID","exprId":{"product-class":"org.apache.spark.sql.catalyst.expressions.ExprId","id":19,"jvmId":"99d99999-9f9f-9afd-99e9-9999c97a9999"},"qualifier":[],"explicitMetadata":{},"nonInheritableMetadataKeys":[]},{"class":"org.apache.spark.sql.catalyst.expressions.Cast","num-children":1,"child":0,"dataType":"integer","timeZoneId":"Etc/UTC"},{"class":"org.apache.spark.sql.catalyst.expressions.AttributeReference","num-children":0,"name":"VENDORID","dataType":"string","nullable":true,"metadata":{},"exprId":{"product-class":"org.apache.spark.sql.catalyst.expressions.ExprId","id":0,"jvmId":"99d99999-9f9f-9afd-99e9-9999c97a9999"},"qualifier":[]}],[{"class":"org.apache.spark.sql.catalyst.expressions.AttributeReference","num-children":0,"name":"TOTALPASSENGER","dataType":"integer","nullable":true,"metadata":{},"exprId":{"product-class":"org.apache.spark.sql.catalyst.expressions.ExprId","id":1,"jvmId":"99d99999-9f9f-9afd-99e9-9999c97a9999"},"qualifier":[]}],[{"class":"org.apache.spark.sql.catalyst.expressions.AttributeReference","num-children":0,"name":"TOTALFARE","dataType":"double","nullable":true,"metadata":{},"exprId":{"product-class":"org.apache.spark.sql.catalyst.expressions.ExprId","id":2,"jvmId":"99d99999-9f9f-9afd-99e9-9999c97a9999"},"qualifier":[]}]],"child":0},{"class":"org.apache.spark.sql.execution.datasources.LogicalRelation","num-children":0,"relation":null,"output":[[{"class":"org.apache.spark.sql.catalyst.expressions.AttributeReference","num-children":0,"name":"VENDORID","dataType":"string","nullable":true,"metadata":{},"exprId":{"product-class":"org.apache.spark.sql.catalyst.expressions.ExprId","id":0,"jvmId":"99d99999-9f9f-9afd-99e9-9999c97a9999"},"qualifier":[]}],[{"class":"org.apache.spark.sql.catalyst.expressions.AttributeReference","num-children":0,"name":"TOTALPASSENGER","dataType":"integer","nullable":true,"metadata":{},"exprId":{"product-class":"org.apache.spark.sql.catalyst.expressions.ExprId","id":1,"jvmId":"99d99999-9f9f-9afd-99e9-9999c97a9999"},"qualifier":[]}],[{"class":"org.apache.spark.sql.catalyst.expressions.AttributeReference","num-children":0,"name":"TOTALFARE","dataType":"double","nullable":true,"metadata":{},"exprId":{"product-class":"org.apache.spark.sql.catalyst.expressions.ExprId","id":2,"jvmId":"99d99999-9f9f-9afd-99e9-9999c97a9999"},"qualifier":[]}]],"isStreaming":false}],"dataSource":null,"options":null,"mode":null}]},"spark_version":{"_producer":"https://github.com/OpenLineage/OpenLineage/tree/0.4.0/integration/spark","_schemaURL":"https://openlineage.io/spec/1-0-2/OpenLineage.json#/$defs/RunFacet","spark-version":"3.1.2","openlineage-spark-version":"0.4.0"}}},"job":{"namespace":"spark_integration","name":"test_synapse_py.execute_save_into_data_source_command"},"inputs":[{"namespace":"sqlserver://synapse-test-test001.sql.azuresynapse.net","name":"dbo.TEST_SUMMARY","facets":{"dataSource":{"_producer":"https://github.com/OpenLineage/OpenLineage/tree/0.4.0/integration/spark","_schemaURL":"https://openlineage.io/spec/facets/1-0-0/DatasourceDatasetFacet.json#/$defs/DatasourceDatasetFacet","name":"sqlserver://synapse-test-test001.sql.azuresynapse.net","uri":"sqlserver://synapse-test-test001.sql.azuresynapse.net"},"schema":{"_producer":"https://github.com/OpenLineage/OpenLineage/tree/0.4.0/integration/spark","_schemaURL":"https://openlineage.io/spec/facets/1-0-0/SchemaDatasetFacet.json#/$defs/SchemaDatasetFacet","fields":[{"name":"VENDORID","type":"string"},{"name":"TOTALPASSENGER","type":"integer"},{"name":"TOTALFARE","type":"double"}]}}}],"outputs":[{"namespace":"sqlserver://synapse-test-test001.sql.azuresynapse.net;databaseName=TESTPOOL1;","name":"dbo.TEST_SUMMARY_COPY_SPARK","facets":{"dataSource":{"_producer":"https://github.com/OpenLineage/OpenLineage/tree/0.4.0/integration/spark","_schemaURL":"https://openlineage.io/spec/facets/1-0-0/DatasourceDatasetFacet.json#/$defs/DatasourceDatasetFacet","name":"sqlserver://synapse-test-test001.sql.azuresynapse.net;databaseName=TESTPOOL1;","uri":"sqlserver://synapse-test-test001.sql.azuresynapse.net;databaseName=TESTPOOL1;"},"schema":{"_producer":"https://github.com/OpenLineage/OpenLineage/tree/0.4.0/integration/spark","_schemaURL":"https://openlineage.io/spec/facets/1-0-0/SchemaDatasetFacet.json#/$defs/SchemaDatasetFacet","fields":[{"name":"VENDORID","type":"integer"},{"name":"TOTALPASSENGER","type":"integer"},{"name":"TOTALFARE","type":"double"}]}},"outputFacets":{"outputStatistics":{"_producer":"https://github.com/OpenLineage/OpenLineage/tree/0.4.0/integration/spark","_schemaURL":"https://openlineage.io/spec/facets/1-0-0/OutputStatisticsOutputDatasetFacet.json#/$defs/OutputStatisticsOutputDatasetFacet","rowCount":0,"size":0}}}],"producer":"https://github.com/OpenLineage/OpenLineage/tree/0.4.0/integration/spark","schemaURL":"https://openlineage.io/spec/1-0-2/OpenLineage.json#/$defs/RunEvent"}
OpenLineageHttpException(code=null, message=<html>
<head>
<meta http-equiv="Content-Type" content="text/html;charset=ISO-8859-1"/>
<title>Error 400 Bad Request</title>
</head>
<body><h2>HTTP ERROR 400 Bad Request</h2>
<table>
<tr><th>URI:</th><td>/api/v1/lineage</td></tr>
<tr><th>STATUS:</th><td>400</td></tr>
<tr><th>MESSAGE:</th><td>Bad Request</td></tr>
<tr><th>SERVLET:</th><td>jersey</td></tr>
</table>

</body>
</html>
, details=null)
        at io.openlineage.spark.agent.OpenLineageContext.emit(OpenLineageContext.java:65)
        at io.openlineage.spark.agent.lifecycle.SparkSQLExecutionContext.endEvent(SparkSQLExecutionContext.java:246)
        at io.openlineage.spark.agent.lifecycle.SparkSQLExecutionContext.end(SparkSQLExecutionContext.java:113)
        at io.openlineage.spark.agent.OpenLineageSparkListener.onJobEnd(OpenLineageSparkListener.java:180)
        at org.apache.spark.scheduler.SparkListenerBus.doPostEvent(SparkListenerBus.scala:39)
        at org.apache.spark.scheduler.SparkListenerBus.doPostEvent$(SparkListenerBus.scala:28)
        at org.apache.spark.scheduler.AsyncEventQueue.doPostEvent(AsyncEventQueue.scala:37)
        at org.apache.spark.scheduler.AsyncEventQueue.doPostEvent(AsyncEventQueue.scala:37)
        at org.apache.spark.util.ListenerBus.postToAll(ListenerBus.scala:117)
        at org.apache.spark.util.ListenerBus.postToAll$(ListenerBus.scala:101)
        at org.apache.spark.scheduler.AsyncEventQueue.super$postToAll(AsyncEventQueue.scala:105)
        at org.apache.spark.scheduler.AsyncEventQueue.$anonfun$dispatch$1(AsyncEventQueue.scala:105)
        at scala.runtime.java8.JFunction0$mcJ$sp.apply(JFunction0$mcJ$sp.java:23)
        at scala.util.DynamicVariable.withValue(DynamicVariable.scala:62)
        at org.apache.spark.scheduler.AsyncEventQueue.org$apache$spark$scheduler$AsyncEventQueue$$dispatch(AsyncEventQueue.scala:100)
        at org.apache.spark.scheduler.AsyncEventQueue$$anon$2.$anonfun$run$1(AsyncEventQueue.scala:96)
        at org.apache.spark.util.Utils$.tryOrStopSparkContext(Utils.scala:1381)
        at org.apache.spark.scheduler.AsyncEventQueue$$anon$2.run(AsyncEventQueue.scala:96)

As you can see, the run event is perfectly created. All inputs ,outputs, schema..everything is fine. Openlineage did create the lineage which should be good to be posted in Marquez, yet it fails to emit the event.

Is there any quick fix available for this issue ?

**Comment by mobuchowski** [comment]
Hey, mind posting the Marquez version and log for error that caused the 400 response? The event looks well-formed and it's valid json, so I guess that Marquez does not like the colons in `sqlserver://synapse-test-test001.sql.azuresynapse.net;databaseName=TESTPOOL1;`

**Reply by Grudra1** [reply]
Hi ! .. Marquez version is 0.19.0. I feel that the colon should not be the problem as for any jdbc dataset/ gcs files the default namespace is the uri for that connection. There we always have a colon and the post works.
Also another point to mention here, in this case not only the lineage post is failed, the marquez becomes pretty much unusable (no namespace box and a black page in loading status forever) untill the services are restarted (for my setup, restart clears the backend tables too so not sure if restart or clearing backend table fixes)

For logs. I have posted above the part where the error starts appearing till the end (post this spark context shutdown hook is called). The html table you are seeing in my above comment is actually part of the log. which got parsed as html while posting here.

Here is the log again :
[error.txt](https://github.com/OpenLineage/OpenLineage/files/8418485/error.txt)

**Reply by Grudra1** [reply]
Tried to post the exact event in lineage api of Marquez..ended up with the same result, may be Openlineage is posting something that Marquez cannot handle ?

![image](https://user-images.githubusercontent.com/102662426/161771977-d1abeeba-a36c-4b4f-b9f9-b2baa97e06e1.png)

**Reply by Grudra1** [reply]
Opened an issue https://github.com/OpenLineage/OpenLineage/issues/655 based on the last comment.
